src/ensembl/doc/architecture/decisions/0016-react-testing-library.md

```diff
@@ -0,0 +1,44 @@
+# 16. React testing library for unit testing React components
+
+Date: 2020-11-28
+
+## Status
+Accepted
+
+## Related ADR(s)
+- [0004-unit-tests](0004-unit-tests.md)
+
+## Context
+
+In [0004-unit-tests](0004-unit-tests.md), we considered two alternative libraries for testing React components: `enzyme` and `react-testing-library`. By that time, we had already been using `enzyme` in `ensembl-client`, and we liked the convenience of selecting components and, if needed, inspecting their props that `enzyme` provided. We therefore decided to stick with `enzyme`. We agreed, however, that our tests would follow the same philosophy as was advocated by the creator of `react-testing-library`, namely that we would never attempt to test implementation details of a component, but would only test component's public API, the html output of a component, and side effects that a component can cause. We thus decided not to use shallow rendering or snapshot testing of serialized components; but rather to test components' behaviour by rendering them fully to the DOM.
+
+Over a year and a half later, it has become clear that the extensive and elegant API of `enzyme`, with its multiple convenience functions for selecting and inspecting components, comes at a significant cost. In order to provide this API, `enzyme` relies on the knowledge of React's internals, which can, and do, change between releases. Every time this happens, `enzyme` has to catch up, and projects relying on `enzyme` need to wait until `enzyme` becomes capable of supporting the latest version of React before being able to update to this version.
```

Comment:
> In order to provide this API, `enzyme` relies on the knowledge of React's internals, which can, and do, change between releases.

This made me wonder if we have other dependencies that are similar to `enzyme` which we need to be concerned of.

Reply:
None that I'm aware of.

The worst offender now is `storybook`, which went so wild with the way it handles its builds that it's incompatible with webpack 5 at the moment. But that's a different story.